src/AggregateFunctions/AggregateFunctionFunnelAnalysis.h

```diff
@@ -0,0 +1,288 @@
+#pragma once
+
+#include <iostream>
+#include <sstream>
+#include <vector>
+#include <unordered_set>
+#include <map>
+#include <Columns/ColumnsNumber.h>
+#include <DataTypes/DataTypeDateTime.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <IO/ReadHelpers.h>
+#include <IO/WriteHelpers.h>
+#include <Columns/ColumnArray.h>
+#include <Columns/ColumnString.h>
+#include <Common/ArenaAllocator.h>
+#include <Common/assert_cast.h>
+#include <common/logger_useful.h>
+#include <Poco/Logger.h>
+
+#include <AggregateFunctions/AggregateFunctionNull.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+}
+
+struct ComparePair final
+{
+    template <typename T1, typename T2>
+    bool operator()(const std::pair<T1, T2> & lhs, const std::pair<T1, T2> & rhs) const
+    {
+        return lhs.first == rhs.first ? lhs.second < rhs.second : lhs.first < rhs.first;
+    }
+};
+
+static constexpr auto max_events = 32;
+template <typename T>
+struct AggregateFunctionFunnelAnalysisData
+{
+    using TimestampEvent = std::pair<T, UInt8>;
+    using TimestampEvents = PODArrayWithStackMemory<TimestampEvent, 64>;
+    using Comparator = ComparePair;
+
+    size_t _size = 0;
+    std::vector<TimestampEvents> events_lists_to_merge;
+
+    size_t size() const
+    {
+        return _size;
+    }
+
+    void add(UInt8 event_idx, const ColumnVector<T> * column, size_t begin, size_t end)
+    {
+        TimestampEvents events;
+        for (size_t i=begin; i<end; i++)
+            events.emplace_back(column->getData()[i], event_idx);
+        events_lists_to_merge.emplace_back(std::begin(events), std::end(events));
```

Comment:
I think it would be better to use `emplace_back(std::move(events))` to avoid extra copying.